src/cascadia/WindowsTerminal/WindowEmperor.cpp

```diff
@@ -18,600 +24,1015 @@ using namespace ::Microsoft::Console;
+        std::unordered_set<winrt::guid> sessionIds;
+        if (const auto layouts = state.PersistedWindowLayouts())
         {
-            for (const auto& actionAndArgs : windowLayout.TabLayout())
+            for (const auto& windowLayout : layouts)
             {
-                const auto args = actionAndArgs.Args();
-                NewTerminalArgs terminalArgs{ nullptr };
-
-                if (const auto tabArgs = args.try_as<NewTabArgs>())
+                for (const auto& actionAndArgs : windowLayout.TabLayout())
                 {
-                    terminalArgs = tabArgs.ContentArgs().try_as<NewTerminalArgs>();
+                    const auto args = actionAndArgs.Args();
+                    NewTerminalArgs terminalArgs{ nullptr };
+
+                    if (const auto tabArgs = args.try_as<NewTabArgs>())
+                    {
+                        terminalArgs = tabArgs.ContentArgs().try_as<NewTerminalArgs>();
+                    }
+                    else if (const auto paneArgs = args.try_as<SplitPaneArgs>())
+                    {
+                        terminalArgs = paneArgs.ContentArgs().try_as<NewTerminalArgs>();
+                    }
+
+                    if (terminalArgs)
+                    {
+                        sessionIds.emplace(terminalArgs.SessionId());
+                    }
                 }
-                else if (const auto paneArgs = args.try_as<SplitPaneArgs>())
+            }
+        }
+
+        // Remove the "buffer_{guid}.txt" files that shouldn't be there
+        // e.g. "buffer_FD40D746-163E-444C-B9B2-6A3EA2B26722.txt"
+        {
+            const std::filesystem::path settingsDirectory{ std::wstring_view{ CascadiaSettings::SettingsDirectory() } };
+            const auto filter = settingsDirectory / L"buffer_*";
+            WIN32_FIND_DATAW ffd;
+
+            // This could also use std::filesystem::directory_iterator.
+            // I was just slightly bothered by how it doesn't have a O(1) .filename()
+            // function, even though the underlying Win32 APIs provide it for free.
+            // Both work fine.
+            const wil::unique_hfind handle{ FindFirstFileExW(filter.c_str(), FindExInfoBasic, &ffd, FindExSearchNameMatch, nullptr, FIND_FIRST_EX_LARGE_FETCH) };
+            if (!handle)
+            {
+                return;
+            }
+
+            do
+            {
+                const auto nameLen = wcsnlen_s(&ffd.cFileName[0], ARRAYSIZE(ffd.cFileName));
+                const std::wstring_view name{ &ffd.cFileName[0], nameLen };
+
+                if (nameLen != 47)
                 {
-                    terminalArgs = paneArgs.ContentArgs().try_as<NewTerminalArgs>();
+                    continue;
                 }
 
-                if (terminalArgs)
+                wchar_t guidStr[39];
+                guidStr[0] = L'{';
+                memcpy(&guidStr[1], name.data() + 7, 36 * sizeof(wchar_t));
+                guidStr[37] = L'}';

```

Comment:
_of course it can't be, that would be too easy_

Reply:
In my defense, when I wrote that `GuidFromPlainString` didn't exist yet. 😅
It's quite ugly code indeed.